src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlBindingPathHelper.cs

```diff
@@ -297,49 +302,67 @@ private static XamlIlBindingPathNode TransformBindingPath(AstTransformationConte
+                            .Where(x => styledElement.IsAssignableFrom(x.Type.GetClrType()))
+                            .Skip(1)
+                            .Where(x => specificAncestorTypeFilter is not null
+                                ? specificAncestorTypeFilter.IsAssignableFrom(x.Type.GetClrType()) : true)
+                            .ElementAtOrDefault(ancestor.Level);
+                        var ancestorType = ancestorNode?.Type.GetClrType();
+
+                        IXamlType? dataContextType = null;
+                        if (ancestorNode is not null)
                         {
-                            var styledElementType = context.GetAvaloniaTypes().StyledElement;
-                            var ancestorType = context
-                                .ParentNodes()
-                                .OfType<XamlAstConstructableObjectNode>()
-                                .Where(x => styledElementType.IsAssignableFrom(x.Type.GetClrType()))
-                                .Skip(1)
-                                .ElementAtOrDefault(ancestor.Level)
-                                ?.Type.GetClrType();
-
-                            if (ancestorType is null)
+                            var isSkipping = true;
+                            foreach (var node in context.ParentNodes())
                             {
-                                throw new XamlX.XamlTransformException("Unable to resolve implicit ancestor type based on XAML tree.", lineInfo);
+                                if (node == ancestorNode)
+                                    isSkipping = false;
+                                if (node is AvaloniaNameScopeRegistrationXamlIlNode)
+                                    break;
+                                if (!isSkipping && node is AvaloniaXamlIlDataContextTypeMetadataNode metadataNode)
+                                {
+                                    dataContextType = metadataNode.DataContextType;
+                                    break;
+                                }
                             }
-
-                            nodes.Add(new FindAncestorPathElementNode(ancestorType, ancestor.Level));
                         }
-                        else
+
+                        if (ancestorType is null)
                         {
-                            nodes.Add(new FindAncestorPathElementNode(GetType(ancestor.Namespace, ancestor.TypeName), ancestor.Level));
+                            if (specificAncestorTypeFilter is not null)
+                                ancestorType = specificAncestorTypeFilter;
+                            else
+                                throw new XamlX.XamlTransformException("Unable to resolve implicit ancestor type based on XAML tree.", lineInfo);
                         }
+
+                        nodes.Add(new FindAncestorPathElementNode(ancestorType, ancestor.Level, dataContextType));
```

Comment:
`ancestorType` is now passed the exact type of the ancestor found, instead of the requested type when it's specified.
This results in a change in behavior:

```xaml
<Border x:Name="p2">
  <Border x:Name="p1">
    <Button x:Name="p0">
      <TextBlock Text="{CompiledBinding $parent[Control;1].Name}" />
    </Button>
  </Border>
</Border>
```

Before this PR, `p1` is found (the 2nd parent `Control`); `ReflectionBinding` does the same.
After this PR, `p2` is found instead (the 2nd parent `Border`).